stats/opentelemetry/internal/grpc_trace_bin_propagator.go

```diff
@@ -0,0 +1,132 @@
+// concerns as both text and binary key-value pairs within a carrier that
+// travels in-band across process boundaries.
+type GRPCTraceBinPropagator struct{}
+
+// Inject set cross-cutting concerns from the Context into the carrier.
+//
+// If carrier is carrier.CustomMapCarrier then SetBinary (fast path) is used,
+// otherwise Set (slow path) with encoding is used.
+func (p GRPCTraceBinPropagator) Inject(ctx context.Context, carrier propagation.TextMapCarrier) {
+	span := trace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return
+	}
+
+	binaryData := Binary(span.SpanContext())
+	if binaryData == nil {
+		return
+	}
+
+	if customCarrier, ok := carrier.(otelinternaltracing.CustomCarrier); ok {
+		customCarrier.SetBinary(binaryData) // fast path: set the binary data without encoding
+	} else {
+		carrier.Set(otelinternaltracing.GRPCTraceBinHeaderKey, base64.StdEncoding.EncodeToString(binaryData)) // slow path: set the binary data with encoding
+	}
+}
+
+// Extract reads cross-cutting concerns from the carrier into a Context.
+//
+// If carrier is carrier.CustomCarrier then GetBinary (fast path) is used,
+// otherwise Get (slow path) with decoding is used.
+func (p GRPCTraceBinPropagator) Extract(ctx context.Context, carrier propagation.TextMapCarrier) context.Context {
+	var binaryData []byte
+
+	if customCarrier, ok := carrier.(otelinternaltracing.CustomCarrier); ok {
+		binaryData, _ = customCarrier.GetBinary()
+	} else {
+		binaryData, _ = base64.StdEncoding.DecodeString(carrier.Get(otelinternaltracing.GRPCTraceBinHeaderKey))
+	}
+	if binaryData == nil {
+		return ctx
+	}
+
+	spanContext, ok := FromBinary([]byte(binaryData))
+	if !ok {
+		return ctx
+	}
+
+	return trace.ContextWithRemoteSpanContext(ctx, spanContext)
+}
+
+// Fields returns the keys whose values are set with Inject.
+//
+// GRPCTraceBinPropagator will only have `grpc-trace-bin` field.
+func (p GRPCTraceBinPropagator) Fields() []string {
+	return []string{otelinternaltracing.GRPCTraceBinHeaderKey}
+}
+
+// Binary returns the binary format representation of a SpanContext.
+//
+// If sc is the zero value, Binary returns nil.
+func Binary(sc trace.SpanContext) []byte {
+	if sc.Equal(trace.SpanContext{}) {
+		return nil
+	}
+	var b [29]byte
+	traceID := trace.TraceID(sc.TraceID())
+	copy(b[2:18], traceID[:])
+	b[18] = 1
+	spanID := trace.SpanID(sc.SpanID())
+	copy(b[19:27], spanID[:])
+	b[27] = 2
+	b[28] = uint8(trace.TraceFlags(sc.TraceFlags()))
+	return b[:]
+}
+
+// FromBinary returns the SpanContext represented by b.
+//
+// If b has an unsupported version ID or contains no TraceID, FromBinary
+// returns with ok==false.
+func FromBinary(b []byte) (sc trace.SpanContext, ok bool) {
+	if len(b) == 0 || b[0] != 0 {
+		return trace.SpanContext{}, false
+	}
+	b = b[1:]
+
+	if len(b) >= 17 && b[0] == 0 {
+		sc = sc.WithTraceID(trace.TraceID(b[1:17]))
+		b = b[17:]
+	} else {
+		return trace.SpanContext{}, false
+	}
```

Comment:
Switch this conditional to early return in the error case as I discussed in a comment above and continue with the operations on line 119 and 120 after the early return.

Reply:
Done